cmd/cosign/cli/verify/verify_blob.go

```diff
@@ -193,220 +164,100 @@ func (c *VerifyBlobCmd) Exec(ctx context.Context, blobRef string) error {
+		if b.Cert != "" {
+			// b.Cert can either be a certificate or public key
+			certBytes := []byte(b.Cert)
+			if isb64(certBytes) {
+				certBytes, _ = base64.StdEncoding.DecodeString(b.Cert)
+			}
+			cert, err = loadCertFromPEM(certBytes)
+			if err != nil {
+				// check if cert is actually a public key
+				co.SigVerifier, err = sigs.LoadPublicKeyRaw(certBytes, crypto.SHA256)
 				if err != nil {
-					return fmt.Errorf("verifying certificate from bundle: %w", err)
+					return fmt.Errorf("loading verifier from bundle: %w", err)
 				}
 			} else {
-				// Verify certificate with chain
-				chain, err := loadCertChainFromFileOrURL(c.CertChain)
-				if err != nil {
-					return err
-				}
-				co.SigVerifier, err = cosign.ValidateAndUnpackCertWithChain(cert, chain, co)
-				if err != nil {
-					return fmt.Errorf("verifying certificate from bundle with chain: %w", err)
+				if c.CertChain != "" {
+					// Load certificate chain
+					chain, err = loadCertChainFromFileOrURL(c.CertChain)
+					if err != nil {
+						return err
+					}
 				}
 			}
 		}
-		if err != nil {
-			return fmt.Errorf("loading verifier from bundle: %w", err)
-		}
-		bundle = b.Bundle
-	default:
-		return fmt.Errorf("please provide a cert to verify against via --certificate or a bundle via --bundle")
-	}
-
-	// Performs all blob verification.
-	if err := verifyBlob(ctx, co, blobBytes, sig, cert, bundle); err != nil {
-		return err
+		opts = append(opts, static.WithBundle(b.Bundle))
 	}
 
-	fmt.Fprintln(os.Stderr, "Verified OK")
-	return nil
-}
-
-/* Verify Blob main entry point. This will perform the following:
-   1. Verifies the signature on the blob using the provided verifier.
-   2. Checks for transparency log entry presence:
-        a. Verifies the Rekor entry in the bundle, if provided. This works offline OR
-        b. If we don't have a Rekor entry retrieved via cert, do an online lookup (assuming
-           we are in experimental mode).
-   3. If a certificate is provided, check it's expiration.
-*/
-// TODO: Make a version of this public. This could be VerifyBlobCmd, but we need to
-// clean up the args into CheckOpts or use KeyOpts here to resolve different KeyOpts.
-func verifyBlob(ctx context.Context, co *cosign.CheckOpts,
-	blobBytes []byte, sig string, cert *x509.Certificate,
-	bundle *bundle.RekorBundle) error {
+	// Gather the cert for the signature.
+	var certPEM []byte
 	if cert != nil {
-		// This would have already be done in the main entrypoint, but do this for robustness.
-		var err error
-		co.SigVerifier, err = cosign.ValidateAndUnpackCert(cert, co)
+		certPEM, err = cryptoutils.MarshalCertificateToPEM(cert)
 		if err != nil {
-			return fmt.Errorf("validating cert: %w", err)
+			return err
+		}
+	}
+	var chainPEM []byte
+	if chain != nil {
+		// Set the last one in the co.RootCerts
+		if co.RootCerts == nil {
+			co.RootCerts = x509.NewCertPool()
+		}
+		co.RootCerts.AddCert(chain[len(chain)-1])
```

Comment:
Does this mean that if we pass a chain, we automatically trust it?

That feels wrong to me.

Reply:
Before the refactoring, this was a way to pass in a chain as an alternative to fetching from the TUF root. Without a chain, we'd default to the TUF metadata, otherwise we'd use the chain passed in.

One bit of feedback is there's a lot of places we're checking `chain != nil` or `c.CertChain != ""`, so it's a little hard to quickly see exactly where we're getting the trust root from.

Reply:
To Zach's point, there are three ways you could pass a root in (this is also all applicable to intermediates):
* SIGSTORE_ROOT_FILE, which will get read in in fulcio.GetRoots
* tuf (either the sigstore tuf root or a self-hosted one), read in in fulcio.GetRoots
* Passing in the cert chain here

I'm not opposed if we want to force the user to set SIGSTORE_ROOT_FILE, as this supports both roots and intermediates (added in https://github.com/sigstore/cosign/pull/1749). I don't recall why this was added this way - likely just a product of iterative development and some overlapping features haha. 

That said, I'm also not opposed to supporting the chain flag too. It really is no different than SIGSTORE_ROOT_FILE, but if it makes things easier, we could drop it. Just wanna make sure we don't lose any supported features.

Reply:
Yeah: this is because in the BLOB case you pass in the cert chain, so you DO trust it "out of band".
In the OCI case, the cert chain is provided on the registry. This was the only way I could get parity on the behavior. the cert chain passed here does represent the full trust chain.

Reply:
> One bit of feedback is there's a lot of places we're checking chain != nil or c.CertChain != "", so it's a little hard to quickly see exactly where we're getting the trust root from.

Just realized you mentioned this so I went and checked. There's only one place that `c.CertChain` is used, and that's to populated `chain` -- after that everything is populated from `chain`. Cleaned that up.

Reply:
LMK if there's anything else here, or we can resolve.

Reply:
Filed https://github.com/sigstore/cosign/issues/2472 as a follow up